### ISSUE TITLE
Improvements to MongoProperties for auto-configuration

### ISSUE THREAD

**Issue by michelzanini** [spring-projects/spring-boot#4017]
Hi,

I am facing one issue while auto-configuring Mongo. My application contains development properties inside the JAR like this:

``` properties
spring.data.mongodb.host=localhost
spring.data.mongodb.port=27017
spring.data.mongodb.database=myDatabase
```

On production, I have a file besides the JAR in the same folder where I wish to override the configuration to connect to a Mongo replica set. Ideally I want it to look like this:

``` properties
spring.data.mongodb.uri=mongodb://myMasterMongoServer:27017,mySlaveMongoServer:27017/myDatabase
```

The problem here is that it will still try to connect to "localhost". That's because since the "host" and "port" properties are in the JAR, and are not overridden, they are still considered.
I would like to propose that the "uri" property takes priority over the "host" and "port", meaning if both are defined together, the "uri" would be the one considered. Like on Rabbit where the "addresses" property prevails over "host".

At first I tried this:

``` properties
spring.data.mongodb.uri=mongodb://myMasterMongoServer:27017,mySlaveMongoServer:27017/myDatabase
spring.data.mongodb.host=
spring.data.mongodb.port=
```

This would set the properties to empty and then I was expecting it to work. However, it still did not work. The reason why is that the "host" property is used if is different then "null", and here it becomes an empty String. I am not aware how we can set a property to "null", I guess it's not possible. So another change could be to check for StringUtils.hasText(host) and not host != null on the class MongoProperties.

In the end, the way I was able to make it work, was to change my property file inside the JAR to also only use the "uri" property like this:

``` properties
spring.data.mongodb.uri=mongodb://localhost:27017/myDatabase
```

Now then production can override the property and host and port are not defined anywhere.
That works, but I still think this could be improved with the suggestions made above.

Thanks.

**Comment by philwebb** [comment]
On the surface this seem like a sensible change but I don't think it's something that we can consider for 1.2 since it would break back compatibility.

@wilkinsona Thoughts for 1.3? or do prefer the status quo?

**Comment by shakuzen** [comment]
Just to add another alternative solution, you could use active profiles (one for dev, one for prod) so that the development environment properties are not even included in the Environment when running on production.

**Comment by snicoll** [comment]
It's inconsistent with Rabbit (and others) so I'd rather have that fixed for 1.3

**Comment by Powerrr** [comment]
And it's also inconsistent with the javadoc - [MongoProperties.java](https://github.com/spring-projects/spring-boot/blob/v1.3.1.RELEASE/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/mongo/MongoProperties.java#L61):
```java
/**
 * Mongo database URI. When set, host and port are ignored.
 */
private String uri = "mongodb://localhost/test";
```
